### ISSUE TITLE
Fail in stop and restart cluster

### ISSUE THREAD

**Issue by MichaelJCompton** [dgraph-io/dgraph#1071]
When I stop individual nodes in a cluster and restart, all seems well.  When I stop all nodes and try to restart the first node brought up crashes.

E.G.

groups.conf
```
default: fp % 3 + 1
```

start servers with

```
~/go/src/github.com/dgraph-io/dgraph/cmd/dgraph/dgraph --group_conf ~/dgraph/groups.conf --groups "0,1" --idx 1 --port 8081 --workerport 12345 --my "127.0.0.1:12345" --debugmode=true --p ~/dgraph/svr1/p --w ~/dgraph/svr1/w > ~/dgraph/svr1/dgraph.log 2>&1 &
```

```
~/go/src/github.com/dgraph-io/dgraph/cmd/dgraph/dgraph --group_conf ~/dgraph/groups.conf --groups "1,2" --idx 2 --port 8082 --workerport 12346 --my "127.0.0.1:12346" --peer "127.0.0.1:12345" --debugmode=true --p ~/dgraph/svr2/p --w ~/dgraph/svr2/w > ~/dgraph/svr2/dgraph.log 2>&1 &
```

```
~/go/src/github.com/dgraph-io/dgraph/cmd/dgraph/dgraph --group_conf ~/dgraph/groups.conf --groups "2,3" --idx 3 --port 8083 --workerport 12347 --my "127.0.0.1:12347" --peer "127.0.0.1:12346" --debugmode=true --p ~/dgraph/svr3/p --w ~/dgraph/svr3/w > ~/dgraph/svr3/dgraph.log 2>&1 &
```

load some data

```
~/go/src/github.com/dgraph-io/dgraph/cmd/dgraphloader/dgraphloader --d 127.0.0.1:8082 --s ~/dgraph/21million.schema --r ~/dgraph/1million.rdf.gz
```

I then waited for the writes on all servers to finish and brought down all the servers with
```
curl -s localhost:8081/admin/shutdown
curl -s localhost:8082/admin/shutdown
curl -s localhost:8083/admin/shutdown
```

When I try to restart, the first node up crashes with

```
2017/06/19 16:38:47 error.go:45: No connection exists

github.com/dgraph-io/dgraph/x.Wrap
	/Users/Michael/go/src/github.com/dgraph-io/dgraph/x/error.go:85
github.com/dgraph-io/dgraph/x.Check
	/Users/Michael/go/src/github.com/dgraph-io/dgraph/x/error.go:45
github.com/dgraph-io/dgraph/worker.(*node).doSendMessage
	/Users/Michael/go/src/github.com/dgraph-io/dgraph/worker/draft.go:409
runtime.goexit
	/usr/local/Cellar/go/1.8.1/libexec/src/runtime/asm_amd64.s:2197
```

**Comment by Rader** [comment]
looks like your memership data corrupt. did you change any `groups` or `ids` when you restart your nodes?

**Comment by MichaelJCompton** [comment]
No, I restart the nodes with everything the same.

**Comment by srh** [comment]
I got the same error, with the following, slightly different procedure (which is the first I tried):
1. Start up the 3 servers as above.
2. Shut them down.
3. Start them up again.
4. Try loading the data as above.  This triggers the error.

**Comment by srh** [comment]
The bug happens because the function `(n *node) doSendMessage(to uint64, data []byte)` in draft.go assumes the following invariant holds:
 - That every addr in `n.peers` has a connection pool entry in `pools()`

The function `func (n *node) Connect(pid uint64, addr string)` tries to maintain this invariant with `connect(addr)` followed by `n.peers.Set(pid, addr)`. That's the only place we add addresses to `n.peers`, so it's a pretty simple invariant to maintain.

The problem is,  `func (p *poolsi) connect(addr string)` can fail to connect, at which point it dumps out `log.Printf("While trying to connect to %q, got error: %v\n", addr, err)` and returns.

So, maybe we should either have `(p *poolsi) connect` return an error, and handle it, or we should do something else.

**Comment by manishrjain** [comment]
Actually, we need a better story about handling connection failures. We don't have a way to act on the health of a peer, for e.g. not sending them queries etc.

**Comment by srh** [comment]
This crash is fixed in master as of 7531a5033798144e2199ef8d04ac6fc0e0ccc112.  I'll leave this open until I make an issue about handling connection failures better (probably tomorrow).
